extensions/amp-analytics/0.1/crypto-impl.js

```diff
@@ -20,29 +20,62 @@ import {getService} from '../../../src/service';
 export class Crypto {
 
   constructor(win) {
-    this.win = win;
+    if (win.crypto) {
+      this.subtle = win.crypto.subtle || win.crypto.webkitSubtle;
+    }
   }
 
   /**
-   * Returns the SHA-384 hash of the input string in an ArrayBuffer.
+   * Returns the SHA-384 hash of the input string in a number array.
+   * Input string cannot contain chars out of range [0,255].
    * @param {string} str
-   * @returns {!Promise<!ArrayBuffer>}
+   * @returns {!Promise<!Array<number>>}
+   * @throws {!Error} when input string contains chars out of range [0,255]
    */
   sha384(str) {
+    if (this.subtle) {
+      try {
+        return this.subtle.digest('SHA-384', str2ab(str))
+            // [].slice.call(Unit8Array) is a shim for Array.from(Unit8Array)
+            .then(buffer => [].slice.call(new Uint8Array(buffer)),
+                () => lib.sha384(str));
+      } catch (e) {
+        // ignore, fallback to closure lib.
+      }
+    }
     return Promise.resolve(lib.sha384(str));
   }
 
   /**
    * Returns the SHA-384 hash of the input string in the format of web safe
-   * base64 string (using -_. instead of +/=).
+   * base64 (using -_. instead of +/=).
+   * Input string cannot contain chars out of range [0,255].
    * @param {string} str
    * @returns {!Promise<string>}
+   * @throws {!Error} when input string contains chars out of range [0,255]
    */
   sha384Base64(str) {
-    return Promise.resolve(lib.sha384Base64(str));
+    return this.sha384(str).then(buffer => {
+      return lib.base64(buffer);
```

Comment:
As I said, the plan is
- Remove the dependency on closure lib for base64 function (might hand write here)
- Exclude base64 code from closure lib, and lazy load the lib only on old browsers

Both things have to happen to gain us benefit. Having an extra implementation here only decrease efficiency and increase code size.

It's out of the scope of this PR though, tracked here: #3690 

On the other hand, we should be able to share the code in someway. Copying the same code here doesn't sound right.